### ISSUE TITLE
Learners should have a representation slot

### ISSUE THREAD

**Issue by pfistfl** [mlr-org/mlr3#445]
I think for many approaches, especially later on in `mlr3pipelines`, we would like to have 
a **representation** slot. 

This is only available for some learners, examples:
- tree based models: Rules for each leaf, and whether obs. belong to this leaf.
- Neural Networks: Output at a given layer (i.e. before softmax)
- Neares Neighbours: NN distance matrix 

Reason: 
This would allow us to extract not only classification but lower-level info from a given learner,
and would, for example, make the implementation of the **[rulefit algorithm](https://christophm.github.io/interpretable-ml-book/rulefit.html)** trivial.

Other examples: https://arxiv.org/abs/1810.09591

Furthermore, we could use this for transfer learning, where representations from a model 
learned on one dataset are transferred to another.

Currently, this would require the learner to expose a `.$representation`  active binding
that calculates this from a given (trained) model.

**Comment by mllg** [comment]
I'm not sure I completely understand what you want here. Is there a definition for a representation?

**Comment by berndbischl** [comment]
these seems impossible to support and implement? every project that tried something similar like this failed, if you now have to talk about / and potentially unify the structure of every possible class of models...

**Comment by pfistfl** [comment]
I will try to rephrase this in order to make it more clear. 
Some learners have an `importance` slot, other's do not.
The same would hold for the `representation`.

A definition of `representation` would be as follows:
In our case, a `representation` for a `Task` of size `n x p` returns a `data.table`  of size `n x q`,
i.e. a row for each original observation.

**Examples** for such a `representation`:
- `rpart`: `factor` that assigns a rule for the leaf node each observation ends up in,
  Ex: `sepal.length > 0.7 && sepal.width < 5`
- `xgboost` | `ranger`: `factor` for each tree, that assigns the rules 
- `neural networks`: Outputs at intermediate layers, e.g. last hidden layer, ncol is then number of 
  neurons in the respective layer.
- `knn`: Distance matrix (if it is stored in the learner). Resulting size would then be `n x n`. 
            (This is more an additional example, as something I would know what to do with)

The goal would be to `cbind` the representation to the original data, as an automatic feature engineering step.

The goal is in no way to unify the structure of every class of models, but instead to use representations learned by different learners as features in other learners. 

I guess one approach would be to write this as a `PipeOp`, that takes a `learner` as input and 
transforms the data into the learner's representation.

**Example of usage**

Something like the `rulefit` algorithm could then be written as follows:

```
gunion(po("extract_representation", lrn("classif.ranger")), po("nop")) %>>%
  po("featureunion") %>>%
  po("glmnet")
```

This yields a fully interpretable `linear model` which incorporates rules. 
Number of rules can be shrunk using L1 - L2 loss on glmnet.

**Comment by mllg** [comment]
Thanks for the explanation. I guess we can have this at some point, but it is rather low priority. If you need this urgent, for your research, you probably have to start with a PR to increase the pressure on us ;)

**Comment by jakob-r** [comment]
What is stopping you from implementing Subclass for the Learner of your choice? You should be able to write a package that implements the specific pipeop and extends some learners with the representation slot. 
Later, if we see that this is an essential part of many learners we could add this slot to the base learner class and have a common definition for how this slot should look like.

**Comment by berndbischl** [comment]
@pfistfl : the problem is this: can you write down a formal definition of what you want? because if you cannot, no one really knows what is in this slot and it cannot be used in general. what you posted above are examples.

**Comment by pfistfl** [comment]
@jakob-r Thanks, that makes sense. Will do that. 
@berndbischl I guess Bengio tried to give a formal definition in this [paper](https://arxiv.org/pdf/1206.5538.pdf).

**Comment by berndbischl** [comment]
> @berndbischl I guess Bengio tried to give a formal definition in this [paper](https://arxiv.org/pdf/1206.5538.pdf).

tbh: on what page exactly does that definition occur? or why not summarize it then here in the thread?

**Comment by pfistfl** [comment]
I guess a super formal definition does not exist, as representations are used in many different contexts.
I.e. a representation of an image can be an image, a tensor, a numeric vector, a character string etc.
analogously, a representation of a tabular observation can also have many different shapes and types.

From the paper:

> representations of the data (that) make it easier to extract useful information when building classifiers or other predictor(s).

>  A good representation is also one that is useful as input to a supervised predictor.

So in order to try to give some definition, I would come up with something like this: 

A representation `R` of a data-point `S` is created via transformation of / feature extraction from the original observation;
 Thus `R = f(S)`, where `f` can encode arbitrary inductive priors, preprocessing steps, etc.
 
The only condition is that a **good** representation  improves the downstream task, e.g. 
training a classifier on  `R` leads to a better performance than training it on `S`.

In the paper, some more priors like `sparsity, smoothness, generalizability` etc. are defined,
not sure if that concerns us though.

**Comment by jakob-r** [comment]
> @jakob-r Thanks, that makes sense. Will do that.

So we can close?

**Comment by berndbischl** [comment]
@pfistfl maybe discuss this orally a bit more with me to get a clearer Pic if you want to

**Comment by mb706** [comment]
> What is stopping you from implementing Subclass for the Learner of your choice? You should be able to write a package that implements the specific pipeop and extends some learners with the representation slot.

To some degree we use inheritance from the base `Learner` because it defines an interface that inheriting Learners should implement. This is useful because then other parts of the code can rely on this common interface. Otherwise we would start implementing some implicit interface by just adding  conceptually similar functions to different concrete `Learner`s, and giving these methods the same name and same signature.

This is a pretty neat idea that should be implemented (and it would play very nicely with mlr3pipelines), I vote for re-opening.
